maitake-sync/src/wait_queue.rs

```diff
@@ -175,7 +177,7 @@ mod tests;
 /// [mutex]: crate::Mutex
 /// [2]: https://www.1024cores.net/home/lock-free-algorithms/queues/intrusive-mpsc-node-based-queue
 #[derive(Debug)]
-pub struct WaitQueue {
+pub struct WaitQueue<Lock: ScopedRawMutex = Spinlock> {
```

Comment:
Misc thought: would it be possible to use a type alias and some feature detection to pick different default locks?

The reason I ask: on embedded, if someone doesn't KNOW to switch out the spinlock, they have a pretty hidden footgun: It will work unless they share stuff with interrupts AND have an interrupt occur while the lock is held.

I can see this being a MAJOR heisenbug for someone to run in to. My suggestion would be to default to the Critical Section mutex (this might be made tricky because it's a feature!) on `target-os = "none"` platforms, and a spinlock on others. Or maybe even not pick a default on those platforms? (this might mean duplicating the struct defn)

I don't know how to make the UX of this not suck, but "it works 99.9% of the time until it doesn't" for embedded platforms is rough.

Reply:
@jamesmunns I have conflicted feelings about this.  There are some competing goals where, unfortunately, I think we have to make some tradeoffs...

In particular, I have a moderate preference for `maitake-sync` not containing or depending on platform-specific code. My goal thus far is that `maitake-sync` just be "plain old Rust" code and any platform-specific behaviors would be provided by the top-level application.

Furthermore, I have a *strong* preference that `maitake-sync` not do its own build-time platform detection. In my experience, this is tricky to get right, and I don't want to be responsible for maintaining it, especially when more robust and more maintained implementations exist elsewhere in the ecosystem. Additionally, there's the issue that sometimes it's not possible to detect a platform without additional user input --- for instance, there are single-core Cortex-M parts, and multi-core Cortex-M parts, and (to my knowledge), Rust can't determine which it's targeting based on just the target spec: instead, the user must set some additional `RUSTFLAGS` cfg or environment variable to indicate it. From what I've seen, there are a bunch of crates that all provide their own mechanism for indicating stuff like this, so users sometimes have to set a bunch of different env vars to indicate to various parts of their dependency graph that they are targeting a single-core platform. I'd really rather not add *another* flag you have to set.

Therefore, I'd much prefer the top-level application build to depend on the crate that does platform detection, and pass its types into `maitake-sync`. Or, if that's unworkable, I'd rather at least have `maitake-sync` depend on some other crate that does the platform detection.  

Regarding `critical-section` support by default, hmm. I'm open to that, but I would really prefer not to make the default mutex type parameter for everything be the `mutex::raw_impls::cs::CriticalSectionRawMutex` type, since it only implements `ScopedRawMutex`, meaning that by default, the RAII mutex APIs won't be available, which seems sad. 

An alternative option could be to bake the `critical_section::acquire` and `critical_section::release` functions into `maitake_sync::blocking::Mutex` itself, rather than the `RawMutex` type. That way, we can always ensure that IRQ disabling and enabling behaviors happen. But, this actually causes a problem when the critical section impl isn't *just* IRQ-disable/IRQ-restore, but also implements a real mutex, because now you have two mutices. I guess `critical-section` does require reentrancy, so it's not a deadlock, it just means you're doing extra stuff... It also kind of makes the `critical-section` impl in `mutex` unnecessary with `maitake-sync`, which is not the end of the world, but we might have to document that. I'd be open to consider this approach, as long as the `critical-section` dep is feature-flagged.

What do you think?

Reply:
For the record, I definitely would prefer to avoid footgun-y default behavior. However, I also kind of feel like platform-specific behaviors like this are usually also *application-specific* behavior, and it's hard to implement any default that really does what you want without *some* additional input from the end user. I dunno. 

Another concern that I forgot to mention above is that I wanted to see if I could add `mutex-traits` support without a breaking change --- which means that the default `new` constructors need to behave identically to how they did before. They can't depend on `critical-section` in that case, since IIUC `critical-section` will fail to link if the user doesn't depend some critical section implementation. So, using `critical-section` by default would be a breaking change, and not having default parameters would also be a breaking change.

With that said, I'm open to releasing a v0.2 if it's necessary. I just wanted to see if I could get away with _not_ doing that. This is not nearly as strong a preference as my desire to avoid having to maintain my own platform detection code!